docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,485 @@
+  // This gives plugins a chance to decide if the request should be allowed to
+  // proceed after the object has passed `rest.Storage` handler validation.
+  // ConstraintEnforcers should not have any side effects. If an error is
+  // returned, the request is marked as failed and no changes are
+  // persisted.
+  EnforceConstraints(a Attributes) error
+}
+
+type Finisher interface {
+  // Finish (name TBD) is executed at the very end of the request lifecycle,
+  // after calls to the `rest.Storage` persistence methods. If an error is
+  // returned, the request is marked as failed, but any changes are preserved
+  // and nothing is rolled back.
+  Finish(a Attributes) error
+}
+
+type Interface interface {
+  // Handles returns true if this admission controller can handle the given
+  // operation where operation can be one of CREATE, UPDATE, DELETE, or CONNECT.
+  Handles(operation Operation) bool
+}
+```
+
+### rest Storage
+
+To be able to invoke admission plugin validation at the right time, we'll need
+to modify the various `rest.Storage` interfaces, as described below.
+
+#### Creater, NamedCreater
+
+- add `BeforeCreate(ctx api.Context, obj runtime.Object) error`
+- add `Validate(ctx api.Context, obj runtime.Object) error`
+
+#### Updater, CreaterUpdater
+
+- add `BeforeUpdate(ctx api.Context, obj, old runtime.Object) error`
+- add `ValidateUpdate(ctx api.Context, obj, old runtime.Object) field.ErrorList`
+
+#### Patcher
+
+- inherits changes from `Updater`
+
+
+#### Deleter, GracefulDeleter
+
+- add `BeforeDelete(ct api.Context, obj runtime.Object, options *api.DeleteOptions) (graceful, gracefulPending bool, err error)`
+
+#### Connector
+
+- QUESTION: do we need to adjust this interface?
+
+
+## REST flows
+
+### Create
+
+#### Current behavior
+
+This is the current [apiserver create
+flow](https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/apiserver/resthandler.go#L333):
+
+1. decode incoming request
+1. admit.Admit
+1. r.Create (the items below are part of the generic registry store / strategy-based storage, `pkg/registry/generic/registry/store.go`)
+  1. rest.BeforeCreate
+    1. strategy.PrepareForCreate
+    1. api.FillObjectMetaSystemFields
+    1. api.GenerateName
+    1. strategy.Validate
+    1. validation.ValidateObjectMeta
+    1. strategy.Canonicalize
+  1. e.Storage.Create
+  1. e.AfterCreate
+  1. e.Decorator
+
+#### Proposed behavior
+
+This is the proposed new behavior:
+
+1. decode request
+1. admit.ApplyDefaults
+1. r.BeforeCreate
+  1. If strategy-based, call rest.BeforeCreate
+    1. strategy.PrepareForCreate
+    1. api.FillInObjectMetaSystemFields
+    1. api.GenerateName
+  1. Otherwise, just do whatever is in the rest.Creater
+1. r.Validate
+  1. If strategy-based, call rest.Validate
+    1. strategy.Validate
+    1. validation.ValidateObjectMeta
+    1. strategy.Canonicalize
+  1. Otherwise, just do whatever is in the rest.Creater
+1. admit.Validate
```

Comment:
Yes, it is definitely not "simple", and the current single-method `Admit` is not really flexible enough. I'm not sure there's an in-between, though.